core/terminal.go

```diff
@@ -52,41 +29,73 @@ type TerminalArgs struct {
 	InsecureRootCapabilities *bool `default:"false"`
 }
 
-func (container *Container) Terminal(svcID *call.ID, args *TerminalArgs) (*Terminal, http.Handler, error) {
-	termID := svcID.Digest()
-	endpoint := "terminals/" + termID.Encoded()
-	term := &Terminal{Endpoint: endpoint}
-	return term, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		var upgrader = websocket.Upgrader{}
-		ws, err := upgrader.Upgrade(w, r, nil)
+func (container *Container) Terminal(
+	ctx context.Context,
+	svcID *call.ID,
+	args *TerminalArgs,
+) error {
+	container = container.Clone()
+
+	container, err := container.WithExec(ctx, ContainerExecOpts{
+		Args:                          args.Cmd,
+		SkipEntrypoint:                true,
+		ExperimentalPrivilegedNesting: *args.ExperimentalPrivilegedNesting,
+		InsecureRootCapabilities:      *args.InsecureRootCapabilities,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create container for interactive terminal: %w", err)
+	}
+
+	return container.terminal(ctx, svcID, args)
+}
+
+func (dir *Directory) Terminal(
+	ctx context.Context,
+	svcID *call.ID,
+	ctr *Container,
+	args *TerminalArgs,
+) error {
+	var err error
+
+	if ctr == nil {
+		ctr, err = NewContainer(dir.Query, dir.Platform)
 		if err != nil {
-			bklog.G(r.Context()).WithError(err).Error("terminal handler failed to upgrade")
-			w.WriteHeader(http.StatusInternalServerError)
-			return
+			return err
 		}
-		defer ws.Close()
+		ctr, err = ctr.From(ctx, "alpine:3.20")
```

Comment:
We have a const for the current version of alpine images we use here: https://github.com/sipsma/dagger/blob/22deba615a453987f70f0f90c9e97f19d0cf18a2/ci/consts/versions.go#L20-L20

Could move that const to `engine/distconsts/consts.go` so then it can be shared here and in our CI module. Just in the interest of keeping everything up to date easily.